app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java

```diff
@@ -501,7 +580,64 @@ private void queueDownload(Download download) {
+                    query.setFilterById(mDownloadReference);
+                    Cursor cursor = manager.query(query);
+                    if (cursor.moveToFirst()) {
+                        int columnIndex = cursor.getColumnIndex(DownloadManager.COLUMN_STATUS);
+                        if (DownloadManager.STATUS_SUCCESSFUL == cursor.getInt(columnIndex)) {
+                            String uriString = cursor.getString(cursor.getColumnIndex(DownloadManager.COLUMN_LOCAL_URI));
+                            cursor.close();
+
+                            if (uriString.substring(0, 7).matches("file://")) {
+                                uriString = uriString.substring(7);
+                            }
+
+                            final String localUri = uriString;
+                            final String fileExtension = MimeTypeMap.getFileExtensionFromUrl(localUri);
+                            final String mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(fileExtension);
+
+                            File file = new File(localUri);
+                            final Uri uriFile = FileProvider.getUriForFile(getContext(), BuildConfig.APPLICATION_ID + ".fileprovider", file);
+
+                            Snackbar snackbar = Snackbar
+                                    .make(browserContainer, String.format(getString(R.string.download_snackbar_finished), fileName), Snackbar.LENGTH_LONG)
+                                    .setAction(getString(R.string.download_snackbar_open), new View.OnClickListener() {
+                                        @Override
+                                        public void onClick(View view) {
+                                            if (mimeType != null) {
+                                                Intent openFileIntent = new Intent(Intent.ACTION_VIEW);
+                                                openFileIntent.setDataAndType(uriFile, mimeType);
+                                                openFileIntent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
+                                                try {
+                                                    startActivity(openFileIntent);
+                                                } catch (ActivityNotFoundException e) {
+                                                    // Ignore if no activity was found.
+                                                }
+                                            }
+                                        }
+                                    });
+                            snackbar.setActionTextColor(ContextCompat.getColor(getContext(), R.color.download_dialog_text));
+                            snackbar.show();
+                        }
+                    } else {
+                        cursor.close();
+                    }
+                }
+            }
+        };
+
+        getActivity().registerReceiver(downloadBroadcastReceiver, filter);
```

Comment:
You register the receiver but you never unregister it (leak). The receiver can be independent from queuing a download. So we could create it once in onCreate() and then register/unregister in onResume() and onPause().